### PR TITLE
Add `process_withdrawals` tests for Gloas

### PR DESCRIPTION
**Pull request by leolara**
Following this previous work:
 + https://github.com/ethereum/consensus-specs/pull/4810
 + https://github.com/ethereum/consensus-specs/pull/4468
 
 This has been rebased to adapt the tests to https://github.com/ethereum/consensus-specs/pull/4788 which changes EIP 7732 making builders to "Make builders non-validating staked actors".

### CODE REVIEW

**Comment by jtraglia** [comment]
Hey @leolara we're planning to make this spec change. When this is merged, please ensure there's at least one test case which fails if these edge cases is not handled properly.

* https://github.com/ethereum/consensus-specs/pull/4832

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md [hunk 1]
```diff
@@ -0,0 +1,214 @@
+# Report: `process_withdrawals` Input/Output Space
+
+Note: This is AI generated from the specs, and it is not the authoritative
+source of truth.
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md]:
```suggestion
*Note*: This report is AI-generated and is not an authoritative source of
truth.
```

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+## Function Signature
+
+```python
+def process_withdrawals(state: BeaconState) -> None
+```
+
+Location:
+[specs/gloas/beacon-chain.md:1029-1055](../../../../../../../specs/gloas/beacon-chain.md#L1029-L1055)
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md]: We should not include line numbers here. It will quickly become outdated & misleading.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+| `state.next_withdrawal_validator_index`                  | `ValidatorIndex`                                                    | `uint64` in \[0, `len(validators)`-1\]. **Bound to** `len(validators)`: wraps via modulo.                                                                            | Starting validator sweep position       |
+| `state.next_withdrawal_builder_index`                    | `BuilderIndex`                                                      | `uint64` in \[0, `len(builders)`-1\]. **Bound to** `len(builders)`: wraps via modulo.                                                                                | Starting builder sweep position         |
+| `state.builders`                                         | `List[Builder, BUILDER_REGISTRY_LIMIT]`                             | Length in [0, 2^20]. Separate non-validating staked actors.                                                                                                          | Builder registry                        |
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md]: Wrong. Where did it get this from?
```suggestion
| `state.builders`                                         | `List[Builder, BUILDER_REGISTRY_LIMIT]`                             | Length in [0, 2^40]. Separate non-validating staked actors.                                                                                                          | Builder registry                        |
```

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+## Key Constants
+
+| Constant                                     | Value                 | Description                           |
+| -------------------------------------------- | --------------------- | ------------------------------------- |
+| `MAX_WITHDRAWALS_PER_PAYLOAD`                | 16                    | Max withdrawals per execution payload |
+| `MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP`       | 16,384                | Max validators checked per sweep      |
+| `MAX_BUILDERS_PER_WITHDRAWALS_SWEEP`         | 16,384                | Max builders checked per sweep        |
+| `MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP` | 8                     | Max partial withdrawals processed     |
+| `VALIDATOR_REGISTRY_LIMIT`                   | 2^40                  | Max validators in registry            |
+| `BUILDER_REGISTRY_LIMIT`                     | 2^20                  | Max builders in registry              |
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md]: Wrong, the builder registration limit is 2**40:
https://jtraglia.github.io/eth-spec-viewer/#v1.7.0-alpha.1/preset_vars-BUILDER_REGISTRY_LIMIT

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+| (in `pending_partial_withdrawals`) `.validator_index`    | `ValidatorIndex`                                                    | `uint64` in \[0, `len(validators)`-1\]. **Index into** `validators[]` and `balances[]`.                                                                              | Identify validator                      |
+| (in `pending_partial_withdrawals`) `.amount`             | `Gwei`                                                              | `uint64`. Requested amount. Actual = `min(balance - MIN_ACTIVATION_BALANCE, amount)`.                                                                                | Withdrawal amount                       |
+| `state.balances[*]`                                      | `Gwei`                                                              | `uint64` in [0, 2^64-1]. **Same length as** `validators[]`. Must have excess over `MIN_ACTIVATION_BALANCE` (32 ETH) for non-slashed withdrawals.                     | Check available balance                 |
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md]: > Must have excess over `MIN_ACTIVATION_BALANCE` (32 ETH) for non-slashed withdrawals.

Not true & misleading in multiple ways:
* It can be less than the min activation balance. There can be an inactive validator with 2 ETH and their balance will still be stored there.
* There does not need to be an "excess". You could deposit 32 ETH and never actually operate a validator. It would take years for it to be ejected.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+   - Builder pending withdrawals: processed first, up to
+     `MAX_WITHDRAWALS_PER_PAYLOAD`
+   - Partial withdrawals: processed second, capped by
+     `MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP`
+   - Builder sweep: processed third, capped by
+     `MAX_BUILDERS_PER_WITHDRAWALS_SWEEP`
+   - Validator sweep: processed last, capped by
+     `MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP`
+   - Total output: `MAX_WITHDRAWALS_PER_PAYLOAD`
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md:121]: This section will need to be updated now that this PR is merged:
* https://github.com/ethereum/consensus-specs/pull/4832

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+- Builders use `BuilderIndex` type, which is converted to/from `ValidatorIndex`
+  using `convert_builder_index_to_validator_index()` and
+  `convert_validator_index_to_builder_index()`
+- In `Withdrawal` output, builder withdrawals use a converted validator index
+  (with `BUILDER_INDEX_FLAG` bit set)
+
+**Builder withdrawal types**:
+
+1. **Builder Pending Withdrawals**: Explicit withdrawal requests in
+   `state.builder_pending_withdrawals`. Uses `fee_recipient` as destination.
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md]: This is misleading. Builders cannot do withdrawal requests. These are payments from builders to proposers.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py [hunk 8]
```diff
@@ -0,0 +1,1324 @@
...
+
+
+def run_gloas_withdrawals_processing(spec, state):
+    """
+    Minimal test harness for process_withdrawals that generates vectors.
+    """
+
+    yield "pre", state
+    spec.process_withdrawals(state)
+    yield "post", state
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py:20]: In the future, we need to consolidate this into the existing `run_withdrawals_processing` function:

* https://github.com/ethereum/consensus-specs/blob/a5a56a99503958b44cb8365e12ed54b204a2bfe4/tests/core/pyspec/eth2spec/test/helpers/withdrawals.py#L206

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py [hunk 9]
```diff
@@ -0,0 +1,1324 @@
...
+    """
+    Test processing when no withdrawals are expected.
+
+    Input State Configured:
+        - Default state with no pending withdrawals configured
+        - latest_block_hash == latest_execution_payload_bid.block_hash (parent block full)
+
+    Output State Verified:
+        - payload_expected_withdrawals: Empty list (len=0)
+        - balances[*]: Unchanged
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py]: It doesn't actually verify this btw.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py [hunk 10]
```diff
@@ -0,0 +1,1324 @@
...
+        - balances[2,3]: == max_effective_balance (partial withdrawals)
+        - next_withdrawal_index: Incremented by 4
+    """
+    fully_withdrawable_indices = [0, 1]
+    partial_withdrawals_indices = [2, 3]
+    all_validator_indices = fully_withdrawable_indices + partial_withdrawals_indices
+    assert max(all_validator_indices) < len(state.validators), (
+        f"Max validator index {max(all_validator_indices)} must exist in registry "
+        f"(validators: {len(state.validators)})"
+    )
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py]: For tests like this, we really don't need these asserts. The test will fail if the value is out-of-bounds.

File: tests/infra/helpers/withdrawals.py [hunk 11]
```diff
@@ -1,11 +1,479 @@
...
+        builder_indices: List of builder indices (BuilderIndex) to add builder pending withdrawals for.
+                        Builders must already exist in state.builders registry.
+        pending_partial_indices: List of validator indices to set up with pending partial withdrawals
+        full_withdrawal_indices: List of validator indices to set up as fully withdrawable
+        partial_withdrawal_indices: List of validator indices to set up as partially withdrawable
+        compounding_indices: List of validator indices to set up with compounding credentials (0x02).
+                            By default, balance is set to MAX_EFFECTIVE_BALANCE_ELECTRA + 1_000_000_000
+                            unless compounding_excess_balances is specified.
+        builder_withdrawal_amounts: Single amount or dict[BuilderIndex, Gwei] for builder withdrawals
+                                    (default: MIN_ACTIVATION_BALANCE for each)
```
Comment (jtraglia) [inline tests/infra/helpers/withdrawals.py]: Why default to 32 ETH? That seems excessive.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py [hunk 12]
```diff
@@ -0,0 +1,1324 @@
...
+
+    Output State Verified:
+        - payload_expected_withdrawals: Contains 1 withdrawal (capped to available balance)
+        - builders[0].balance: 0 (all available balance withdrawn)
+        - builder_pending_withdrawals: Reduced by 1 (processed even if capped)
+        - next_withdrawal_index: Incremented by 1
+    """
+    builder_index = 0
+    withdrawal_amount = spec.Gwei(5_000_000_000)
+    available_balance = spec.Gwei(1_000_000_000)
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py:129]: I would like to do some bounds testing here. A builder with this balance would never be able to withdrawal. It would be better to set these to something like:
```suggestion
    withdrawal_amount = spec.MIN_DEPOSIT_AMOUNT + spec.Gwei(123)
    available_balance = spec.MIN_DEPOSIT_AMOUNT + spec.Gwei(122)
```

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py [hunk 13]
```diff
@@ -0,0 +1,1324 @@
...
+
+    assert_process_withdrawals(
+        spec,
+        state,
+        pre_state,
+        withdrawal_count=1,
+        builder_balances={builder_index: 0},
+        builder_pending_delta=-1,
+        withdrawal_index_delta=1,
+    )
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py]: This is an impossible situation. You cannot have an active builder with less than `MIN_DEPOSIT_AMOUNT` balance. I'm not sure this test case should exist. Its redeeming quality is that it ensures there cannot be negative builder balances.

Edit: Just keep it, we appear to have other "impossible situation" tests elsewhere.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py [hunk 14]
```diff
@@ -0,0 +1,1324 @@
...
+        - next_withdrawal_index: Incremented by processed count
+    """
+
+    pending_indices = list(range(spec.MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP))
+    assert max(pending_indices) < len(state.validators), (
+        f"Max validator index {max(pending_indices)} must exist in registry "
+        f"(validators: {len(state.validators)})"
+    )
+
+    prepare_process_withdrawals(spec, state, pending_partial_indices=pending_indices)
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py]: It would be good to have _more_ than `MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP` pending partial withdrawals here, to ensure that it's limited properly.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py [hunk 15]
```diff
@@ -0,0 +1,1324 @@
...
+        partial_withdrawal_indices=[validator_index],
+        validator_exit_epoch_offsets={validator_index: 2},
+    )
+
+    assert spec.is_active_validator(
+        state.validators[validator_index], spec.get_current_epoch(state)
+    )
+    assert not spec.is_active_validator(
+        state.validators[validator_index], spec.get_current_epoch(state) + 2
+    )
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py]: Why +2 here instead of +1?

Pre-gloas it's this:
* https://github.com/ethereum/consensus-specs/blob/a5a56a99503958b44cb8365e12ed54b204a2bfe4/tests/core/pyspec/eth2spec/test/capella/block_processing/test_process_withdrawals.py#L768-L773

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py:1]: A lot of the test cases here are duplicates of tests from other forks, see:

* https://github.com/ethereum/consensus-specs/blob/master/tests/core/pyspec/eth2spec/test/capella/block_processing/test_process_withdrawals.py
* https://github.com/ethereum/consensus-specs/blob/master/tests/core/pyspec/eth2spec/test/electra/block_processing/test_process_withdrawals.py

Some but not all are `@with_all_phases_from_to(CAPELLA, GLOAS)`. Something to keep in mind. I would like to delete the duplicate tests where possible. Ideally, this file only contains tests for code changed by Gloas.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py (hunk 15 above)
Comment (jtraglia) [reply tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py]: Since this is a duplicate test, we can just delete this.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py:1]: There are no builder sweep withdrawal tests...

**Review by jtraglia — CHANGES_REQUESTED** [review]
Here's my initial quick review. There's a lot going on here 😅 For the next round of changes, let's remove duplicate tests for pre-Gloas functionality; the tests in this PR should focus on stuff that Gloas changes. Another big thing I noticed is that there are no builder sweep withdrawal tests. Let's look into adding those afterwards.

**Comment by leolara** [comment]
@jtraglia About the duplicates:

There are duplicates. I **think** that all were already in the first try to this PR by terence: https://github.com/ethereum/consensus-specs/pull/4468

I tried to make sure I kept all the of those. I thought that because of the block parent thing, they needed to be different in Gloas (perhaps because of the parent block thing). Perhaps, this was true at some point and that is the reason they are there  and in capella says `@with_all_phases_from_to(CAPELLA, GLOAS)`.

I will check all of them.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md (hunk 6 above)
Comment (leolara) [reply tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.md:121]: Bad timing :-) I guess that will also require to change the last test.

File: tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py (hunk 8 above)
Comment (leolara) [reply tests/core/pyspec/eth2spec/test/gloas/block_processing/test_process_withdrawals.py:20]: I think that would require to change more the previous versions of the test and I didn't want to do that. But it is possible.

**Comment by leolara** [comment]
I will take it out of draft once it is compatible with merged: https://github.com/ethereum/consensus-specs/pull/4832

**Comment by leolara** [comment]
I have:

  - Fix some things in the test documentation
  - Removed duplicate tests from capella that came with PR #4468
  - Added test_builder_withdrawal_insufficient_balance_realistic_bounds for bounds testing
  - Added test_pending_withdrawals_processing_exceeds_limit for limit enforcement
  - Added 4 builder sweep withdrawal tests
  - Added tests for validators and builders not in the registry
  - Added edge case test for PR #4832 (test_full_builder_payload_reserves_sweep_slot)
  - Changed default builder withdrawal amount from 32 ETH to 1 ETH
  - Moved registry existence checks to prepare_process_withdrawals
  - Documented that the bug is now fixed
  
  This has been merged with master that has PR #4832

**Comment by leolara** [comment]
It passes with preset=mainnet

**Review by jtraglia — APPROVED** [review]
LGTM. I have skimmed the new tests and they all look reasonable to me. Let's merge this now & iterate if/when we identify more edge cases. Thank you @leolara!

**Comment by jihoonsong** [comment]
The `test_add_builder_updates_existing` test in `test_builders.py` checks if `add_builder_to_registry` updates existing builder correctly, but `add_builder_to_registry` in the spec always creates new builder. I think this test can be deleted. cc @leolara
